test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MvcEndpointDataSourceTests.cs

```diff
@@ -152,16 +152,22 @@ private static TheoryData GetSingleActionData(bool isConventionalRouting)
                     {"{controller}/{id?}", isConventionalRouting ? new string[] { } : new[] { "TestController/{id?}" }},
                     {"{action}/{id?}", isConventionalRouting ? new string[] { } : new[] { "TestAction/{id?}" }},
                     {"{Controller}/{Action}/{id?}", new[] { "TestController/TestAction/{id?}" }},
+                    {"{Controller}/{Action}/{id?}/{more?}", new[] { "TestController/TestAction/{id?}/{more?}" }},
                     {"{CONTROLLER}/{ACTION}/{id?}", new[] { "TestController/TestAction/{id?}" }},
                     {"{controller}/{action=TestAction}", new[] { "TestController", "TestController/TestAction" }},
-                    {"{controller}/{action=TestAction}/{id?}", new[] { "TestController", "TestController/TestAction/{id?}" }},
-                    {"{controller=TestController}/{action=TestAction}/{id?}", new[] { "", "TestController", "TestController/TestAction/{id?}" }},
+                    {"{controller}/{action=TestAction}/{id?}", new[] { "TestController/TestAction/{id}", "TestController", "TestController/TestAction" }},
```

Comment:
Another thing to consider here is that a parameter with a default value behaves a lot like an optional parameter.

So like `{controller}/{action=TestAction}/{id=17}` - what's the link generation behavior of that route?

Reply:
We got a bug here, both matching and generating. I'm going to play around with it but if you see this before it is fixed and you have 10 minutes free, ping me, I'd like to ask some Qs.

Reply:
`{controller=Home}/{action=Index}/{id=17}` with `HomeController.Index(string id)`

Ye old routing:
Matching - `/` succeeds and id is 17
Generation - `/` is output from `@Url.Action("Index", "Home", new { id = "17" })`

Endpoint routing:
Matching - `/` fails. `/Home/Index` succeeds and id is 17
Generation - `/Home/Index/17` is output from `@Url.Action("Index", "Home", new { id = "17" })`

What we could do when denormalizing endpoints is treat a parameter with a default as optional and included additional endpoints with shortened patterns, but we need a way to specify that "17" is still the value of `id`. Does such a thing exist today? If not should it be added, or do you have a better idea?

Reply:
I've fix matching but generation outputs `/Home/Index` with `@Url.Action("Index", "Home", new { id = "17" })`. I'm pretty sure the issue is link generation doesn't know to not output `/Home/Index` because they are just literal parameters now.

I think we'll have to include endpoints with matching suppressed to get around it.

Reply:
> I was thinking about this at the gym. If there is a matching suppressed endpoint with the parameters still present then the generated results could include non-endpoint URLs again.

Wouldn't required values/keys shut this down?